### ISSUE TITLE
LLVM errors

### ISSUE THREAD

**Issue by leonardo-m** [AdamNiederer/faster#1]
I am seeing errors with debug compilation:

```
...>cargo build
   Compiling stdsimd v0.0.3
   Compiling faster v0.1.3 (file:.../faster_master)
LLVM ERROR: Cannot select: t7: v16i8 = X86ISD::ABS t3
  t3: v16i8,ch = CopyFromReg t0, Register:v16i8 %vreg45
    t2: v16i8 = Register %vreg45
In function: _ZN6faster4main28_$u7b$$u7b$closure$u7d$$u7d$17h9a8972f3fea441f3E
error: Could not compile `faster`.
```

And release compilation:

```
...>cargo build --release
   Compiling faster v0.1.3 (file:.../faster_master)
LLVM ERROR: Cannot select: t144: v16i8 = X86ISD::ABS t158
  t158: v16i8 = bitcast t157
    t157: v2i64,ch = load<LD16[ConstantPool]> t0, t160, undef:i64
      t160: i64 = X86ISD::WrapperRIP TargetConstantPool:i64<<16 x i8> <i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10>> 0
        t159: i64 = TargetConstantPool<<16 x i8> <i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10, i8 -10>> 0
      t12: i64 = undef
In function: _ZN6faster4main17hb45e5f709eed73c2E
error: Could not compile `faster`.
```

The compiler is the last Nightly:

```
rustc 1.23.0-nightly (3b82e4c74 2017-11-05)
binary: rustc
commit-hash: 3b82e4c74d43fc1273244532c3a90bf9912061cf
commit-date: 2017-11-05
host: x86_64-pc-windows-gnu
release: 1.23.0-nightly
LLVM version: 4.0
```

**Comment by AdamNiederer** [comment]
Hello,
This looks like an LLVM issue. Could you tell me what platform you're compiling on (and for), and what feature level you're compiling for (native, x86-64, ivybridge...)?

**Comment by leonardo-m** [comment]
Both compilations are just for x86-64. The OS is a Win 8.1.

**Comment by GabrielMajeri** [comment]
@AdamNiederer I can confirm that building this crate on Linux x64 with `-C target-cpu=native` works, but when I remove the `-C target-cpu` flag I also get compilation errors.

Perhaps the code for plain x86_64 does not know what intrinsics to use? If you're compiling for x86_64, does `rustc` automatically enable the feature `target_feature = "sse"` or does that have to be enabled? Because it looks like SSE is supported but LLVM fails to generate SSE code.

**Comment by AdamNiederer** [comment]
Ah, thank you for the repro. I've isolated the problem to the `as_i8s` call in `main`. There's also an issue related to _mm_round_ps, but that's definitely in stdsimd (might be because my working branch is dirty).

I'll see what I can do about the transmutation functions; those are high on the list anyway because they have a chance to cause UB.

**Comment by AdamNiederer** [comment]
All of these issues were resolved in 0.1.4 - the transmutation methods were changed from `as_*` to `be_*`, and the floating point transmutation methods are marked as unsafe for now.

The root of the issue was that x86_64 isn't guaranteed to have SSE4.1, which was causing my `abs()` polyfill to emit an intrinsic which isn't supported on bare `x86_64`.
